pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaBytesSource.java

```diff
@@ -41,16 +49,80 @@
+
     @Override
     protected Properties beforeCreateConsumer(Properties props) {
-        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
-        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        props.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        props.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
         log.info("Created kafka consumer config : {}", props);
+
+        String currentKeyDeserializer = props.getProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG);
+        if (!SUPPORTED_KEY_DESERIALIZERS.contains(currentKeyDeserializer)) {
+            throw new IllegalArgumentException("Unsupported key deserializer: " + currentKeyDeserializer + ", only " + SUPPORTED_KEY_DESERIALIZERS);
+        }
+
+        String currentValueDeserializer = props.getProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG);
+        if (!SUPPORTED_VALUE_DESERIALIZERS.contains(currentValueDeserializer)) {
+            throw new IllegalArgumentException("Unsupported value deserializer: " + currentValueDeserializer + ", only " + SUPPORTED_VALUE_DESERIALIZERS);
+        }
+
+        // replace KafkaAvroDeserializer with our custom implementation
+        if (currentValueDeserializer != null && currentValueDeserializer.equals(KafkaAvroDeserializer.class.getName())) {
+            props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, SchemaExtractorDeserializer.class.getName());
+            KafkaAvroDeserializerConfig config = new KafkaAvroDeserializerConfig(props);
+            List<String> urls = config.getSchemaRegistryUrls();
+            int maxSchemaObject = config.getMaxSchemasPerSubject();
+            SchemaRegistryClient schemaRegistryClient = new CachedSchemaRegistryClient(urls, maxSchemaObject);
+            schemaCache = new AvroSchemaCache(schemaRegistryClient);
+        }
         return props;
     }
 
     @Override
-    public byte[] extractValue(ConsumerRecord<String, byte[]> record) {
-        return record.value();
+    public Object extractValue(ConsumerRecord<Object, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithSchema) {
+            return ((BytesWithSchema) value).getValue();
+        }
+        return value;
     }
+
+    @Override
+    public org.apache.pulsar.client.api.Schema<byte[]> extractSchema(ConsumerRecord<Object, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithSchema) {
+            return schemaCache.get(((BytesWithSchema) value).getSchemaId());
+        } else {
+            return org.apache.pulsar.client.api.Schema.BYTES;
+        }
+    }
+
+    public static class SchemaExtractorDeserializer implements Deserializer<BytesWithSchema> {
```

Comment:
Can you rename it as ExtractKafkaAvroSchemaDeserializer? Because this is specific to Kafka avro schema.

Reply:
done